specifications/cloud_assembly.md

```diff
@@ -0,0 +1,473 @@
+* The *Cloud Assembly Specification* is extensible.
+* The *Cloud Assembly Specification* is cloud-agnostic.
+* The *Cloud Assembly Specification* is easy to implement and use.
+* The *Cloud Assembly Specification* supports authenticity and integrity guarantees.
+* A *Cloud Assembly* is self-contained, making deployments reproductible.
+
+## Specification
+A *Cloud Assembly* is a ZIP archive that **SHOULD** conform to the [ISO/IEC 21320-1:2015] *Document Container File*
+standard. *Cloud Assembly* files **SHOULD** use the `.cloud` extension in order to make them easier to recognize by
+users.
+
+Documents in the archive can be stored with any name and directory structure, however the following entries at the root
+of the archive are reserved for special use:
+* `manifest.json` **MUST** be present and contains the [manifest document](#manifest-document) for the *Cloud Assembly*.
+* `signature.asc`, when present, **MUST** contain the [digital signature](#digital-signature) of the *Cloud Assembly*.
+
+### Manifest Document
+The `manifest.json` file is the entry point of the *Cloud Assembly*. It **MUST** be a valid [JSON] document composed of
+a single `object` that conforms to the following schema:
+
+Key           |Type                 |Required|Description
+--------------|---------------------|:------:|-----------
+`schema`      |`string`             |Required|The schema for the document. **MUST** be `cloud-assembly/1.0`.
+`drops`       |`Map<ID,Drop>`       |Required|A mapping of [*Logical ID*](#logical-id) to [Drop](#drop).
+`missing`     |`Map<string,Missing>`|        |A mapping of context keys to [missing information](#missing).
+
+The [JSON] specification allows for keys to be specified multiple times in a given `object`. However, *Cloud Assembly*
+consumers **MAY** assume keys are unique, and *Cloud Assemblers* **SHOULD** avoid generating duplicate keys. If
+duplicate keys are present and the manifest parser permits it, the latest specified value **SHOULD** be preferred.
+
+### Logical ID
+*Logical IDs* are `string`s that uniquely identify [Drop](#drop)s in the context of a *Cloud Assembly*.
+* A *Logical ID* **MUST NOT** be empty.
+* A *Logical ID* **SHOULD NOT** exceed `256` characters.
+* A *Logical ID* **MUST** be composed of only the following ASCII printable characters:
+  + Upper-case letters: `A` (`0x41`) through `Z` (`0x5A`)
+  + Lower-case letters: `a` (`0x61`) through `z` (`0x7A`)
+  + Numeric characters: `0` (`0x30`) through `9` (`0x39`)
+  + Plus: `+` (`0x2B`)
+  + Minus: `-` (`0x2D`)
+  + Forward-slash: `/` (`0x2F`)
+  + Underscore: `_` (`0x5F`)
+* A *Logical ID* **MUST NOT** contain the `.` (`0x2E`) character as it is used in the string substitution pattern for
+  cross-drop references to separate the *Logical ID* from the *attribute* name.
+
+In other words, *Logical IDs* are expected to match the following regular expression:
+```js
+/^[A-Za-z0-9+\/_-]{1,256}$/
```

Comment:
Can it be things like `-`, `+`, `_`, `0` then? Seems a little too permissive maybe?

Reply:
Why? I have no problem with those. Their primary goal is not human-readability, it's supposed to be machine-processed. I find expressing "only letters and numbers, -, +, _, but must start by a letter" starts looking like the expression of a poor password policy... I don't want to impose restrictions that are convoluted to express in plain English unless I have a good reason to.

Reply:
I don't think `0`, or `-`, or `+` are great logical IDs, but maybe it's just me. Feel free to ignore.